src/Compilers/VisualBasic/Portable/SymbolDisplay/SymbolDisplay.vb

```diff
@@ -115,30 +115,121 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         ''' <remarks>
         ''' Handles <see cref="Boolean"/>, <see cref="String"/>, <see cref="Char"/>, <see cref="SByte"/>
         ''' <see cref="Byte"/>, <see cref="Short"/>, <see cref="UShort"/>, <see cref="Integer"/>, <see cref="UInteger"/>,
         ''' <see cref="Long"/>, <see cref="ULong"/>, <see cref="Double"/>, <see cref="Single"/>, <see cref="Decimal"/>,
-        ''' <see cref="Date"/>, and <c>Nothing</c>.
+        ''' <see cref="Date"/>, and <see langword="Nothing"/>.
         ''' </remarks>
         Public Function FormatPrimitive(obj As Object, quoteStrings As Boolean, useHexadecimalNumbers As Boolean) As String
-            Dim options = ObjectDisplayOptions.None
-            If quoteStrings Then
-                options = options Or ObjectDisplayOptions.UseQuotes Or ObjectDisplayOptions.EscapeNonPrintableCharacters
+            Return ObjectDisplay.FormatPrimitive(obj, ToObjectDisplayOptions(quoteStrings, useHexadecimalNumbers))
+        End Function
+
+        ''' <summary>
+        ''' Returns a textual representation of an object of primitive type as an array of string parts,
+        ''' each of which has a kind. Useful for colorizing the display string.
+        ''' </summary>
+        ''' <param name="type">The type of the value.</param>
+        ''' <param name="value">A value to display as string parts.</param>
+        ''' <param name="format">The formatting options to apply. If <see langword="Nothing"/> is passed, <see cref="SymbolDisplayFormat.VisualBasicErrorMessageFormat"/> will be used.</param>
+        ''' <returns>A list of display parts (or <see langword="Nothing"/> if the type is not supported).</returns>
+        ''' <remarks>
+        ''' Handles <see cref="Boolean"/>, <see cref="String"/>, <see cref="Char"/>, <see cref="SByte"/>
+        ''' <see cref="Byte"/>, <see cref="Short"/>, <see cref="UShort"/>, <see cref="Integer"/>, <see cref="UInteger"/>,
+        ''' <see cref="Long"/>, <see cref="ULong"/>, <see cref="Double"/>, <see cref="Single"/>, <see cref="Decimal"/>,
+        ''' <see cref="Date"/>, and <see langword="Nothing"/>.
+        ''' </remarks>
+        Public Function PrimitiveToDisplayParts(type As ITypeSymbol, value As Object, Optional format As SymbolDisplayFormat = Nothing) As ImmutableArray(Of SymbolDisplayPart)
+            If Not (value Is Nothing OrElse value.GetType().IsPrimitive OrElse TypeOf value Is String OrElse TypeOf value Is Decimal OrElse TypeOf value Is Date) Then
+                Return Nothing
             End If
-            If useHexadecimalNumbers Then
-                options = options Or ObjectDisplayOptions.UseHexadecimalNumbers
+
+            Dim builder = ArrayBuilder(Of SymbolDisplayPart).GetInstance()
+            If type Is Nothing Then
+                AddConstantValue(builder, value, ToObjectDisplayOptions(If(format, SymbolDisplayFormat.VisualBasicErrorMessageFormat).ConstantValueOptions))
```

Comment:
>If(format, SymbolDisplayFormat.VisualBasicErrorMessageFormat) [](start = 72, length = 61)

Would it make sense to de-dupe this? Both branches calculate this unconditionally.